src/main/resources/lib/login.js

```diff
@@ -4,67 +4,38 @@ const configLib = require('/lib/config');
-
-    //Retrieves the user
+function login(token, tokenClaims, isAutoLogin) {
     const idProviderKey = portalLib.getIdProviderKey();
-    const userName = commonLib.sanitize(preconditions.checkParameter(userinfoClaims, 'sub'));
+    const idProviderConfig = configLib.getIdProviderConfig();
+    const userName = commonLib.sanitize(preconditions.checkParameter(tokenClaims, idProviderConfig.claimUsername));
     const principalKey = 'user:' + idProviderKey + ':' + userName;
     const user = contextLib.runAsSu(() => authLib.getPrincipal(principalKey));
 
-    //If the user does not exist
-    if (!user) {
-
-        //Creates the users
-        const idProviderConfig = configLib.getIdProviderConfig();
-
-        if (idProviderConfig.rules.forceEmailVerification) {
-            preconditions.check(userinfoClaims.email_verified === true, 'Email must be verified');
+    try {
+        let claims = tokenClaims;
+        if (!user) {
+            if (!isAutoLogin || idProviderConfig.autoLogin.createUsers) {
+                claims = resolveClaims(idProviderConfig, token, tokenClaims);
+                doCreateUser(idProviderConfig, claims.userinfo, userName, isAutoLogin);
+            } else if (isAutoLogin) {
+                throwAutoLoginFailedError(`Auto login failed for user '${userName}'. User does not exist`);
+            }
         }
 
-        const email = idProviderConfig.mappings.email.replace(regExp, (match, claimKey) => getClaim(claims, claimKey)) || null;
-        const displayName = idProviderConfig.mappings.displayName.replace(regExp, (match, claimKey) => getClaim(claims, claimKey)) ||
-                            userinfoClaims.preferred_username || userinfoClaims.name || email || userinfoClaims.sub;
-
-        const user = contextLib.runAsSu(() => authLib.createUser({
-            idProvider: idProviderKey,
-            name: userName,
-            displayName: displayName,
-            email: email
-        }));
-        log.info(`User [${user.key}] created`);
-
-        contextLib.runAsSu(() => {
-            idProviderConfig.defaultGroups.forEach(function (defaultGroup) {
-                authLib.addMembers(defaultGroup, [user.key]);
-                log.debug(`User [${user.key}] added to group [${defaultGroup}]`);
-            });
-        });
-    }
-
-    //Updates the profile
-    const profile = contextLib.runAsSu(() => authLib.modifyProfile({
-        key: principalKey,
-        scope: 'oidc',
-        editor: () => removeNonSupportedKeys(claims)
-    }));
-    log.debug(`Modified profile of [${principalKey}]: ${JSON.stringify(profile)}`);
-
-    //Logs in the user
-    const loginResult = authLib.login({
-        user: userName,
-        idProvider: idProviderKey,
-        skipAuth: true
-    });
-    if (loginResult.authenticated) {
-        log.debug(`Logged in user [${principalKey}]`);
-    } else {
-        throw `Error while logging user [${principalKey}]`;
+        modifyUserProfile(claims, principalKey);
```

Comment:
do we always try to modify a user on every login? This should not be done at least for autologin

Reply:
Are you sure? `modifyUserProfile` removes non supported claims.